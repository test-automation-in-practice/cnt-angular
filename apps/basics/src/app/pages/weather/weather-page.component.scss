@import '../../../../../basics/src/styles/page';
