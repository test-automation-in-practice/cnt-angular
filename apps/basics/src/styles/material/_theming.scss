@use '@angular/material' as mat;

@import '../colors';

@include mat.core();

$angular-typography: mat.define-typography-config(
  $font-family: 'Roboto, "Helvetica Neue", sans-serif',
  $headline-1: mat.define-typography-level(24px, 38px, 400),
  $headline-2: mat.define-typography-level(20px, 34px, 400),
  $subtitle-1: mat.define-typography-level(20px, 34px, 400),
  $subtitle-2: mat.define-typography-level(18px, 30px, 400),
  $body-2: mat.define-typography-level(16px, 26px, 400),
  $body-1: mat.define-typography-level(16px, 26px, 400),
  $caption: mat.define-typography-level(inherit, 1.5, 400),
  $button: mat.define-typography-level(inherit, 1, 400),
);

$ntred: (
  100: var(--palette-accent-100),
  200: var(--palette-accent-200),
  300: var(--palette-accent-300),
  400: var(--palette-accent-400),
  500: var(--palette-accent-500),
  600: var(--palette-accent-600),
  700: var(--palette-accent-700),
  800: var(--palette-accent-800),
  900: var(--palette-accent-900),
  A100: var(--palette-accent-a100),
  A200: var(--palette-accent-a200),
  A400: var(--palette-accent-a400),
  A700: var(--palette-accent-a700),
  contrast: (
    50: #000000,
    100: #000000,
    200: #000000,
    300: #000000,
    400: #000000,
    500: #ffffff,
    600: #ffffff,
    700: #ffffff,
    800: #ffffff,
    900: #ffffff,
    A100: #000000,
    A200: #000000,
    A400: #000000,
    A700: #000000,
  ),
);

$ntviolette: (
  50: var(--palette-primary-50),
  100: var(--palette-primary-100),
  200: var(--palette-primary-200),
  300: var(--palette-primary-300),
  400: var(--palette-primary-400),
  500: var(--palette-primary-500),
  600: var(--palette-primary-600),
  700: var(--palette-primary-700),
  800: var(--palette-primary-800),
  900: var(--palette-primary-900),
  A100: var(--palette-primary-a100),
  A200: var(--palette-primary-a200),
  A400: var(--palette-primary-a400),
  A700: var(--palette-primary-a700),
  contrast: (
    50: #000000,
    100: #000000,
    200: #000000,
    300: #000000,
    400: #ffffff,
    500: #ffffff,
    600: #ffffff,
    700: #ffffff,
    800: #ffffff,
    900: #ffffff,
    A100: #000000,
    A200: #000000,
    A400: #ffffff,
    A700: #ffffff,
  ),
);

$angular-primary: mat.define-palette($ntviolette, 500, 100, 900);
$angular-accent: mat.define-palette($ntred, 500, 100, 900);

$angular-warn: mat.define-palette(mat.$red-palette);

$angular-default-theme: mat.define-dark-theme(
  (
    color: (
      primary: $angular-primary,
      accent: $angular-accent,
    ),
  )
);

@include mat.all-component-themes($angular-default-theme);

//@include mat.expansion-theme()
