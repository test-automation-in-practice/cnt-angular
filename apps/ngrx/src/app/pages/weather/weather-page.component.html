<h1>Weather App Basic examples</h1>
<cntws-weather-introduction class="block pb-4"></cntws-weather-introduction>
<div class="block relative min-h-100">
  <ng-container *ngIf="(isLoading$ | async) !== true">
    <cntws-location
      *ngIf="(mainLocation$ | async) !== null"
      (searched)="search($event)"
      [defaultLocation]="(mainLocation$ | async)!"
      class="block pb-4"
    ></cntws-location>
    <cntws-weather-results
      *ngIf="(weather$ | async)?.length ?? 0 > 0"
      [weatherLocations]="(weather$ | async)!"
      (savedLocation)="saveLocation($event)"
      class="block pb-4"
    ></cntws-weather-results>
    <cntws-error-message *ngIf="warning$ | async as message" [message]="message!" class="block mb-3"></cntws-error-message>
  </ng-container>
  <cntws-loading *ngIf="isLoading$ | async"></cntws-loading>
</div>
<cntws-weather-explanation class="block pb-4"></cntws-weather-explanation>
