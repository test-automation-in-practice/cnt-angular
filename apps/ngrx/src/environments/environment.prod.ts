export const environment = {
  production: true,
  weatherApi: 'http://localhost:4300',
};
