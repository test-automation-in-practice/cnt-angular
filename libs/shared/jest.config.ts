/* eslint-disable */
const esModules = ['@angular', '@ngrx'];

export default {
  displayName: 'shared',
  preset: '../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {},
  coverageDirectory: '../../coverage/libs/shared',
  transform: {
    '^.+\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: [`<rootDir>/node_modules/(?!.*\\.mjs$|${esModules.join('|')})`],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
