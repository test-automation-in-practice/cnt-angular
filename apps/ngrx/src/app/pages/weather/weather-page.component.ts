import { Component } from "@angular/core";
import { WeatherLocation } from "@cntws/weather";
import { WeatherFacade } from "./+state/weather.facade";
import { map } from 'rxjs';

@Component({
  selector: 'cntws-weather-page',
  templateUrl: './weather-page.component.html',
  styleUrls: ['./weather-page.component.scss'],
})
export class WeatherPageComponent {
  weather$ = this.weather.weather$;
  warning$ = this.weather.error$;
  isLoading$ = this.weather.loaded$.pipe(map(l => !l));
  mainLocation$ = this.weather.mainLocation$;

  constructor(private weather: WeatherFacade) {
    this.weather.init();
  }

  search(location: string) {
    this.weather.loadWeatherForLocation(location);
  }

  saveLocation(location: WeatherLocation) {
    this.weather.saveMainLocation(location.location);
  }
}
