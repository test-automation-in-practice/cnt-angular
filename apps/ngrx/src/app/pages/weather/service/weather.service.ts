import { Injectable } from "@angular/core";
import { environment } from "../../../../environments/environment";
import { HttpClient } from "@angular/common/http";
import { map, Observable } from "rxjs";
import { LocationApiModel, WeatherApiModel, WeatherLocation } from '@cntws/weather';

@Injectable()
export class WeatherService {
  static ENDPOINTS = {
    mainLocation: () => `${environment.weatherApi}/mainLocation`,
    queryLocation: (location: string) => `${environment.weatherApi}/locations?q=${location}`,
  };

  constructor(private http: HttpClient) {}

  private static mapApiModel(apiModel: WeatherApiModel[]): WeatherLocation[] {
    return apiModel.map((apiWeather) => ({ temp: apiWeather.temp, location: apiWeather.name, id: apiWeather.id }));
  }

  getMainLocation(): Observable<string> {
    return this.http
      .get<LocationApiModel>(WeatherService.ENDPOINTS.mainLocation())
      .pipe(map(loc => loc.name));
  }

  getWeatherForLocation(location: string): Observable<WeatherLocation[]> {
    return this.http
      .get<WeatherApiModel[]>(WeatherService.ENDPOINTS.queryLocation(location))
      .pipe(map(WeatherService.mapApiModel));
  }

  saveMainLocation(location: string): Observable<string> {
    return this.http
      .post<LocationApiModel>(WeatherService.ENDPOINTS.mainLocation(), { name: location })
      .pipe(map(loc => loc.name));
  }
}
