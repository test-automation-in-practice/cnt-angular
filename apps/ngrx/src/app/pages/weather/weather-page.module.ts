import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WeatherPageComponent } from './weather-page.component';
import { WeatherService } from './+state/service/weather.service';
import { ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatInputModule } from '@angular/material/input';
import { MatCardModule } from '@angular/material/card';
import { MatDividerModule } from '@angular/material/divider';
import { WeatherIntroductionComponent } from './paragraphs/introduction/weather-introduction.component';
import { WeatherExplanationComponent } from './paragraphs/explanation/weather-explanation.component';
import { MatIconModule } from '@angular/material/icon';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatExpansionModule } from '@angular/material/expansion';
import { WeatherModule } from '@cntws/weather';
import { ErrorMessageComponent, LoadingComponent } from '@cntws/shared';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromWeather from './+state/weather.reducer';
import { WeatherEffects } from './+state/weather.effects';
import { WeatherFacade } from './+state/weather.facade';

@NgModule({
  declarations: [WeatherPageComponent, WeatherIntroductionComponent, WeatherExplanationComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatButtonModule,
    MatInputModule,
    MatCardModule,
    MatDividerModule,
    MatIconModule,
    MatProgressSpinnerModule,
    MatExpansionModule,
    WeatherModule,
    ErrorMessageComponent,
    LoadingComponent,
    StoreModule.forFeature(fromWeather.WEATHER_FEATURE_KEY, fromWeather.weatherReducer),
    EffectsModule.forFeature([WeatherEffects]),
  ],
  providers: [WeatherService, WeatherFacade],
})
export class WeatherPageModule {}
