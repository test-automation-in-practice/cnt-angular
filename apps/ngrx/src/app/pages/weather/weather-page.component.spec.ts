import '../../testing-helpers/window.mock';
import { MockBuilder, MockRender, MockService, ngMocks } from 'ng-mocks';
import { WeatherPageComponent } from './weather-page.component';
import { WeatherPageModule } from './weather-page.module';
import { WeatherService } from './+state/service/weather.service';
import { of } from 'rxjs';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { environment } from '../../../environments/environment';
import { MatCardHarness } from '@angular/material/card/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { MatInputHarness } from '@angular/material/input/testing';
import { MatButtonHarness } from '@angular/material/button/testing';
import { MatCardModule } from '@angular/material/card';
import { WeatherIntroductionComponent } from './paragraphs/introduction/weather-introduction.component';
import { WeatherExplanationComponent } from './paragraphs/explanation/weather-explanation.component';
import { MatExpansionModule } from '@angular/material/expansion';
import { MatInputModule } from '@angular/material/input';
import { MatButtonModule } from '@angular/material/button';
import { ReactiveFormsModule } from '@angular/forms';
import { MatDividerModule } from '@angular/material/divider';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { WeatherLocation, WeatherModule } from '@cntws/weather';
import { ErrorMessageComponent, LoadingComponent } from '@cntws/shared';
import { StoreModule } from '@ngrx/store';
import * as fromWeather from './+state/weather.reducer';
import { EffectsModule } from '@ngrx/effects';
import { WeatherEffects } from './+state/weather.effects';
import { WeatherFacade } from './+state/weather.facade';
import { ApiModelGenerators } from '@cntws/testing';

type serviceMockProps = {
  isLoading?: boolean;
  warning?: string | undefined;
  weather?: WeatherLocation[];
  mainLocation?: string | undefined;
};

function createWeatherFacade(opts: serviceMockProps) {
  const stateOverrides = {
    loaded$: of(!opts.isLoading),
    error$: of(opts.warning),
    weather$: of(opts.weather ? opts.weather : []),
    mainLocation$: of(opts.mainLocation),
  };
  return MockService(WeatherFacade, stateOverrides);
}

const MaterialModules = [MatCardModule, MatExpansionModule, MatInputModule, MatButtonModule, MatDividerModule, MatProgressSpinnerModule];

describe('Weather Page: A user visiting the page', () => {
  /**
   * These tests are testing the complete module and give us confidence for bigger refactorings.
   * They are rather brittle and thus, we have to use them carefully.
   * Further, they do not give good insights in why a specific scenario is failing.
   * This leads to a need of additional tests for your components to get better feedback.
   */
  describe('and having their data loaded', () => {
    let fixture: ComponentFixture<WeatherPageComponent>;
    let loader: HarnessLoader;
    let controller: HttpTestingController;

    beforeEach(async () => {
      return TestBed.configureTestingModule({
        declarations: [WeatherPageComponent, WeatherIntroductionComponent, WeatherExplanationComponent],
        imports: [
          NoopAnimationsModule,
          WeatherModule,
          HttpClientTestingModule,
          ReactiveFormsModule,
          ErrorMessageComponent,
          LoadingComponent,
          ...MaterialModules,
          StoreModule.forRoot({ [fromWeather.WEATHER_FEATURE_KEY]: fromWeather.weatherReducer }),
          EffectsModule.forRoot([WeatherEffects]),
        ],
        providers: [WeatherService, WeatherFacade],
      }).compileComponents();
    });

    beforeEach(() => {
      fixture = TestBed.createComponent(WeatherPageComponent);
      loader = TestbedHarnessEnvironment.loader(fixture);
      controller = TestBed.inject(HttpTestingController);
    });

    it('should load their saved location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      await loader.getHarness(MatCardHarness.with({ subtitle: new RegExp(`.*(${location}).*`) }));
    });

    it('should load the weather for their saved location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      await loader.getHarness(MatCardHarness.with({ title: location }));
    });

    it('should be able to search for a location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      const locationCard = await loader.getHarness(MatCardHarness.with({ subtitle: new RegExp(`.*(location).*`) }));
      const locationInput = await locationCard.getHarness(MatInputHarness);
      const searchedWeather = 'Frankfurt';
      await locationInput.setValue(searchedWeather);
      const searchButton = await locationCard.getHarness(MatButtonHarness);
      await searchButton.click();
      const searchedWeatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${searchedWeather}`);
      searchedWeatherRequest.flush(ApiModelGenerators.createWeatherApiModel(searchedWeather, temp));

      await loader.getHarness(MatCardHarness.with({ title: searchedWeather }));
    });
  });

  describe('should have a consistent layout', () => {
    beforeEach(() => {
      ngMocks.reset();
      return MockBuilder(WeatherPageComponent, WeatherPageModule);
    });

    it('if the service is loading', () => {
      const facade = createWeatherFacade({ isLoading: true });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherFacade, useValue: facade }] });
      expect(fixture.point.nativeElement).toMatchSnapshot();
    });

    it('if the service has a warning', () => {
      const facade = createWeatherFacade({ warning: 'This is a warning!' });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherFacade, useValue: facade }], reset: true });
      expect(fixture.point.nativeElement).toMatchSnapshot();
    });

    it('if the service has no warning and results', () => {
      const facade = createWeatherFacade({
        weather: [
          { id: 1, location: 'Stuttgart', temp: 23 },
          { id: 2, location: 'Frankfurt', temp: 15 },
        ],
      });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherFacade, useValue: facade }] });
      expect(fixture.point.nativeElement).toMatchSnapshot();
    });
  });
});
