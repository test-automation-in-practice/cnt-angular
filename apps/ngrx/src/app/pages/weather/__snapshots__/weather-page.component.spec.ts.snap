// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Page: A user visiting the page should have a consistent layout if the service has a warning 1`] = `
<cntws-weather-page>
  <h1>
    Weather App Basic examples
  </h1>
  <cntws-weather-introduction
    class="block pb-4"
  />
  <div
    class="block relative min-h-100"
  >
    <cntws-location
      class="block pb-4"
    />
    
    
    <cntws-error-message
      class="block mb-3"
    />
    
    
    
    
  </div>
  <cntws-weather-explanation
    class="block pb-4"
  />
</cntws-weather-page>
`;

exports[`Weather Page: A user visiting the page should have a consistent layout if the service has no warning and results 1`] = `
<cntws-weather-page>
  <h1>
    Weather App Basic examples
  </h1>
  <cntws-weather-introduction
    class="block pb-4"
  />
  <div
    class="block relative min-h-100"
  >
    <cntws-location
      class="block pb-4"
    />
    
    <cntws-weather-results
      class="block pb-4"
    />
    
    
    
    
    
  </div>
  <cntws-weather-explanation
    class="block pb-4"
  />
</cntws-weather-page>
`;

exports[`Weather Page: A user visiting the page should have a consistent layout if the service is loading 1`] = `
<cntws-weather-page>
  <h1>
    Weather App Basic examples
  </h1>
  <cntws-weather-introduction
    class="block pb-4"
  />
  <div
    class="block relative min-h-100"
  >
    
    <cntws-loading />
    
  </div>
  <cntws-weather-explanation
    class="block pb-4"
  />
</cntws-weather-page>
`;
